crates/sui-core/src/execution_scheduler/mod.rs

```diff
@@ -0,0 +1,1115 @@
+                    epoch_store.epoch(),
+                    cert.0.epoch(),
+                );
+                None
+            }
+        });
+
+        for cert in certs {
+            let scheduler = self.clone();
+            let epoch_store = epoch_store.clone();
+            spawn_monitored_task!(
+                epoch_store.within_alive_epoch(scheduler.schedule_transaction(
+                    cert.0,
+                    cert.1,
+                    &epoch_store,
+                ))
+            );
+        }
+
+        self.metrics
+            .transaction_manager_num_pending_certificates
+            .set(self.pending_certificates.read().len() as i64);
+    }
+
+    async fn schedule_transaction(
+        self,
+        cert: VerifiedExecutableTransaction,
+        expected_effects_digest: Option<TransactionEffectsDigest>,
+        epoch_store: &Arc<AuthorityPerEpochStore>,
+    ) {
+        if !self.pending_certificates.write().insert(*cert.digest()) {
+            return;
+        }
+        let digest = cert.digest();
+        tracing::debug!(?digest, "Schedule_transaction");
+
+        self.overload_tracker.add_pending_certificate(cert.data());
+        let enqueue_time = Instant::now();
+        let tx_data = cert.transaction_data();
+        let input_object_kinds = tx_data
+            .input_objects()
+            .expect("input_objects() cannot fail");
+
+        let input_object_keys: Vec<_> =
+            match epoch_store.get_input_object_keys(&cert.key(), &input_object_kinds) {
+                Ok(keys) => keys,
+                Err(_) => {
+                    // This is possible if the transaction is already executed.
+                    // TODO: Eventually we could pass assigned shared object versions
+                    // to the scheduler so that this call cannot return Err.
+                    assert!(self.transaction_cache_read.is_tx_already_executed(digest));
+                    return;
+                }
+            }
+            .into_iter()
+            .collect();
+        let receiving_object_keys: HashSet<_> = tx_data
+            .receiving_objects()
+            .into_iter()
+            .map(|entry| {
+                InputKey::VersionedObject {
+                    // TODO: Add support for receiving ConsensusV2 objects. For now this assumes fastpath.
+                    id: FullObjectID::new(entry.0, None),
+                    version: entry.1,
+                }
+            })
+            .collect();
+
+        let input_and_receiving_keys = [
+            input_object_keys,
+            receiving_object_keys.iter().cloned().collect(),
+        ]
+        .concat();
+
+        let epoch = epoch_store.epoch();
+        let digests = [*digest];
+        let queue_time = Instant::now();
+        tracing::trace!(
+            ?digests,
+            "Waiting for input objects: {:?}",
+            input_and_receiving_keys
+        );
+
+        tokio::select! {
```

Comment:
i think we may want to spawn the waiter task here. By doing all the work above in the task we will probably consume a lot more memory since the Future will have to include space for all the stack variables here.

Reply:
Most of the variables are used below though

Reply:
Updated this